### DISCUSSION TITLE
Can not get UEFI client working

### DISCUSSION THREAD

**Discussion by paykocaydin** [ltsp/ltsp#653] [Q&A]
Hello, 

I have installed LTSP as described in the https://ltsp.org/docs/installation/. It is a 2 NIC setup, LTSP network is configured as 192.168.67.0, 192.168.67.1 being the LTSP server. I have tried both with Ubuntu 20.04 and Debian with the same results. 

The problem is that legacy (non UEFI) clients works fine, UEFI mode/ UEFI clients does not work and stops after DHCPACK message. 
First I have tried with DELL WYSE 5070 diskless thin client which has only the UEFI option. When this did not work i have tried with Lenovo E440 notebook, without a disk. When i put Lenovo's NIC in UEFI mode it does not work but when i put in in legacy mode it works. 

I do not know what is wrong, it seems that in UEFI mode everything is going fine, system identifies the arch. correctly, sends the correct file that the client has to load. Then the communication stops and client says it can not find a bootable source. (or something similar). 

I have attached my ltsp.dnsmasq.conf file. 
Also i have attached 2 log files. 
Legacy is the log of succesful session in legacy mode. 
efi is the log on unsuccesful session in UEFI mode. 

Any help, suggestions will be much appriciated. 

Aydin

[ltsp-dnsmasq.conf.txt](https://github.com/ltsp/ltsp/files/7965856/ltsp-dnsmasq.conf.txt)
[legacy.txt](https://github.com/ltsp/ltsp/files/7965857/legacy.txt)
[efi.txt](https://github.com/ltsp/ltsp/files/7965858/efi.txt)

**Comment by alkisg** [comment]
Try booting a client from a USB stick with ipxe.efi on it, to see if the problem is client side (e.g. incompatible or old UEFI implementation) or server side (e.g. wrong settings). Or try with a VM client in UEFI mode.

**Reply by paykocaydin** [reply]
Thank you Alkis for your suggestions. I have tried with 2 different clients, even changed the switch in the network. I do not think it is a client issue but i will try with a VM client in UEFI mode tomorrow.

Also in my configuration there is no ipxe.efi but there is snponly.efi. 
I will also try replacing snponly.efi with ipxe.efi.

Aydin

**Comment by alkisg** [comment]
You can find ipxe.efi there: http://boot.ipxe.org
Or you can use my liveusb image from there: https://github.com/alkisg/liveusb
Using ipxe.efi will allow us to type `config` in the ipxe prompt and see if the client receives all the necessary information from DHCP, or what exactly is missing.

**Reply by paykocaydin** [reply]
Alkis thank you again for your time and help. 
I have made some progress, i want to describe you what i have done and ask for your opinion again. 

I have replaced /etc/dnsmasq.d/ltsp-dnsmasq.conf with this

port=0
enable-tftp
tftp-root=/srv/tftp
log-dhcp
dhcprange=192.168.67.20,182.168.67.100,13h
dhcp-option=option:dns-server,0.0.0.0
dhcp-no-override

pxe-prompt="Menu"
pxe-service=0, "Boot BIOS", ltsp/undionly.kpxe
pxe-service=7, "Boot UEFI", ltsp/snponly.efi
pxe-service=0, "Exit Menu"
pxe-service=7, "Exit Menu" 

Now when i boot the thin client this menu appears.
It also says "Station IP address is 192.168.67.68" which means it gets the IP from dhcp.
When i select "Boot UEFI" it loads snponly.efi but it stops and displays the message "No more network devices" 

During this booting process i am able to enter ipxe command line by pressing ctrl-b
I can see the config as you have described in your other post, all necessary information seems to be there except filename which i do not know what it does. 
In the command line if i write 

boot ltsp/ltsp.ipxe 

System loads ltsp.ipxe and a menu displays, In a couple of seconds it automatically loads
vmlinuz
ltsp.img
initrd.img 

and the os loads as it should be. 

Regards,

Aydin

**Comment by alkisg** [comment]
I think that your UEFI firmware is borked and it needs special conditions to boot.
E.g. maybe it requires the pxe-prompt line.
Unfortunately I can't really help with that, it's up to you to pinpoint the exact line that is needed.
It will be easier though if you start with the original ltsp-dnsmasq.conf, and just add the lines that you think the client needs, otherwise you'll be missing the appropriate boot filename sequence etc, like you already saw.

**Reply by paykocaydin** [reply]
Hello again, 

I have made changes in ltsp-dnsmasq.conf and added

pxe-service=tag:!ipxe,X86-64_EFI,"snponly.efi",ltsp/snponly.efi

Now it is working but only if there are 2 lines of the same code. (Same line twice) 
If there is only one of them, UEFI client can not boot. 

If there are 2 lines a menu is displayed like 
snponly.efi
snponly.efi 

When you choose one of them system boots normally. 

Can this be a timing issue? 

Aydin

**Reply by paykocaydin** [reply]
Hi Alkis, 

Finally found a solution. 
Actually i found one of your posts related to this issue. 
https://lists.thekelleys.org.uk/pipermail/dnsmasq-discuss/2021q1/014854.html

Commented out all the pxe-service lines and it worked without any problems. 

Regards and thanks again.

Aydin

**Reply by intelliant01** [reply]
Hi, just saw this. You have arrived. :)

The problem has been explained herein - 
https://lists.thekelleys.org.uk/pipermail/dnsmasq-discuss/2021q3/015677.html

The bug is in dnsmasq. The proper solution is to replace the default dnsmasq executable with the new one compiled from the patched branch of Petr - https://github.com/InfrastructureServices/dnsmasq/tree/pxe-services

Then use the default ltsp-dnsmasq.conf and all should work as expected.

**Reply by paykocaydin** [reply]
Hi, 

Thank you for your reply and link for the proper solution. 

Aydin

**Reply by Tuharsky** [reply]
I'm glad to find this thread, and commenting out the 5 lines from dnsmasq conf file changes things a bit, however my client still dosen't boot. Only writes these lines:

NBP filename is ltsp/snponly.efi
NBP filename is 181952 Bytes
Downloading NBP file...
Succeed to download NBP file

and then returns to boot menu.

**Comment by alkisg** [comment]
@Tuharsky try to disable secure boot in the UEFI settings.

**Reply by Tuharsky** [reply]
Unfortunately, this is not possible since it would ruin alternative boot from local drive that is installed as UEFI enabled. This HP Elitebook dosen't allow combination of legacy+UEFI boot.

**Reply by alkisg** [reply]
When disabling secure boot, the PC still boots in UEFI mode.
When enabling legacy boot (CSM) the PC then boots in BIOS mode.

I'm asking to test disabling secure boot. Not to enable legacy boot (CSM).

**Reply by Tuharsky** [reply]
Tested and confirming the same problem with HP Elitedesk 800 DM (miniPC) - impossible to enable legacy boot along with secure boot. This is so already with the oldest G1 model.
Some time ago I encountered this with a higher end Lenovo notebook.
I'm afraid it is going to be increasingly more difficult to avoid UEFI. What can I do to make LTSP boot that way?

**Reply by Tuharsky** [reply]
You are right, disabling Secure boot allowed to enable legacy boot options. That way LTSP is able to boot.
My fault. Thank You, Alkis.
